### PR TITLE
[CPDLP-1580] Access management - Switch between deliver partners

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

- Ticket: https://dfedigital.atlassian.net/browse/CPDLP-1580

### Changes proposed in this pull request

* DP participants page updated to allow for multiple organisations and switching between them
* DP choose organisation page
* Login as `delivery-partner@example.com`
  * If the user only has one DP partner role, it will redirect to the participants
  * If more than one, will redirect to choose organisation page

### CODE REVIEW

File: app/controllers/appropriate_bodies/participants_controller.rb [hunk 1]
```diff
@@ -2,6 +2,16 @@
...
+      appropriate_body
+    end
+
+  private
+
+    def appropriate_body
+      @appropriate_body ||= current_user.appropriate_bodies.find(params[:appropriate_body_id])
+    end
+
+    helper_method :appropriate_body
```
Comment (asmega) [inline app/controllers/appropriate_bodies/participants_controller.rb:15]: how come a change in approach here? I personally prefer this method over `before_action` as they are lazy

File: app/forms/delivery_partners/choose_organisation_form.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      end
+    end
+
+    def delivery_partner
+      return if delivery_partner_id.blank?
+
+      user.delivery_partners.find(delivery_partner_id)
+    end
+
+    def only_one
```
Comment (asmega) [inline app/forms/delivery_partners/choose_organisation_form.rb:25]: predicate?

File: app/views/admin/delivery_partners/users/edit.html.erb [hunk 3]
```diff
@@ -8,13 +8,13 @@
...
 
-    <%= form_with model: @delivery_partners_user_form, scope: :delivery_partners_user_form, url: admin_delivery_partners_user_path(@user) do |f| %>
+    <%= form_with model: @delivery_partners_user_form, scope: :delivery_partners_user_form, url: admin_delivery_partners_user_path(@delivery_partner_profile) do |f| %>
       <%= f.govuk_collection_select(:delivery_partner_id, DeliveryPartner.name_order, :id, :name, label: { text: "Delivery partner" }, options: { include_blank: true }, class: ["js-location-select"]) %>
       <%= f.govuk_text_field(:full_name, label: { text: "Full name" }) %>
       <%= f.govuk_email_field(:email, label: { text: "Email" }) %>
       <div class="govuk-button-group">
         <%= f.govuk_submit "Save" %>
-        <%= govuk_button_link_to "Delete", delete_admin_delivery_partners_user_path(@user), class: "govuk-button--warning", "data-test"=> "delete-button" %>
+        <%= govuk_button_link_to "Delete", delete_admin_delivery_partners_user_path(@delivery_partner_profile), class: "govuk-button--warning", "data-test"=> "delete-button" %>
```
Comment (asmega) [inline app/views/admin/delivery_partners/users/edit.html.erb:17]: RE `data-test`,  personally not a fan of adding cruft to the dom. and pretty sure it can be targeted with a decent selector. is this being used?

File: app/forms/delivery_partners/update_user_form.rb [hunk 4]
```diff
@@ -3,47 +3,46 @@
...
-      @email = user.email
-      @delivery_partner_id = user.delivery_partner_profile&.delivery_partner_id
-    end
+    delegate :user, to: :delivery_partner_profile
 
-    def delivery_partner
-      @delivery_partner ||= DeliveryPartner.find(delivery_partner_id)
+    def initialize(*args)
+      super
+      self.full_name = user.full_name
```
[the remark this answers is not in the document]
Comment (mooktakim) [reply app/forms/delivery_partners/update_user_form.rb:24]: they won't be preloaded from the user model. Only at attribute assignment.
In this case, we want to preload the existing name, email and DP so the edit form can display it.

File: app/controllers/appropriate_bodies/participants_controller.rb (hunk 1 above)
Comment (mooktakim) [reply app/controllers/appropriate_bodies/participants_controller.rb:15]: @asmega 
I guess this is more to stay consistent with what we do in other controllers, we don't use `before_action` much.

File: app/views/admin/delivery_partners/users/edit.html.erb (hunk 3 above)
Comment (mooktakim) [reply app/views/admin/delivery_partners/users/edit.html.erb:17]: @asmega 
I don't think we need `data-test` here. I'll remove it

File: app/forms/delivery_partners/choose_organisation_form.rb (hunk 2 above)
Comment (mooktakim) [reply app/forms/delivery_partners/choose_organisation_form.rb:25]: @asmega 
As we're setting the `delivery_partner_id` here, I've removed the `?`

File: app/controllers/appropriate_bodies/appropriate_bodies_controller.rb [hunk 5]
```diff
@@ -26,6 +24,10 @@ def create
 
   private
 
+    def set_choose_organisation_form
```
Comment (cwrw) [inline app/controllers/appropriate_bodies/appropriate_bodies_controller.rb:27]: instead of using `before_action`, can we try set the attribute memoized?
```
def choose_organisation_form
    @choose_organisation_form ||= ChooseOrganisationForm.new(user: current_user)
end
```

File: app/models/delivery_partner_profile.rb [hunk 6]
```diff
@@ -6,6 +6,8 @@ class DeliveryPartnerProfile < ApplicationRecord
   belongs_to :user
   belongs_to :delivery_partner
 
+  scope :with_users, -> { includes(:user) }
```
Comment (cwrw) [inline app/models/delivery_partner_profile.rb:9]: I wonder if we do need this scope, as we're using it only once 🤔

Comment (mooktakim) [reply app/models/delivery_partner_profile.rb]: @cwrw 
sure, I'll get rid of it

File: app/controllers/admin/delivery_partners/users_controller.rb [hunk 7]
```diff
@@ -5,11 +5,10 @@ module DeliveryPartners
     class UsersController < Admin::BaseController
       skip_after_action :verify_authorized
       skip_after_action :verify_policy_scoped
-      before_action :set_user, only: %i[edit update delete destroy]
+      before_action :set_delivery_partner_profile_and_authorize, only: %i[edit update delete destroy]
 
       def index
-        sorted_users = User.delivery_partner_users.name_order
-        @pagy, @users = pagy(sorted_users, page: params[:page], items: 20)
+        @pagy, @delivery_partner_profiles = pagy(DeliveryPartnerProfile.includes(:user), page: params[:page], items: 20)
```
Comment (cwrw) [inline app/controllers/admin/delivery_partners/users_controller.rb:11]: could we sort by name here as well?

File: app/forms/delivery_partners/choose_organisation_form.rb [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    include ActiveModel::Model
+    include ActiveModel::Attributes
+
+    attribute :user
+    attribute :delivery_partner_id
+
+    validates :delivery_partner_id, inclusion: { in: :valid_ids }
+
+    def delivery_partner_options
+      @delivery_partner_options ||= user.delivery_partners.each_with_object({}) do |ab, sum|
```
Comment (cwrw) [inline app/forms/delivery_partners/choose_organisation_form.rb:14]: ➕

File: app/views/admin/delivery_partners/users/delete.html.erb [hunk 9]
```diff
@@ -1,9 +1,9 @@
 <% content_for :title, "Delete user" %>
 <h1>Do you want to delete this user?</h1>
-<p class="govuk-!-font-weight-bold">Supplier user: <span class="govuk-!-font-weight-regular"><%= @user.full_name %></span></p>
```
Comment (cwrw) [inline app/views/admin/delivery_partners/users/delete.html.erb:3]: I think we are missing the summary error message box in this form

File: app/views/admin/delivery_partners/users/edit.html.erb [hunk 10]
```diff
@@ -8,13 +8,13 @@
   <div class="govuk-grid-column-two-thirds">
     <h1 class="govuk-heading-l">Edit user details</h1>
 
-    <%= form_with model: @delivery_partners_user_form, scope: :delivery_partners_user_form, url: admin_delivery_partners_user_path(@user) do |f| %>
```
Comment (cwrw) [inline app/views/admin/delivery_partners/users/edit.html.erb:11]: same here regarding missing the summary error box (not part of this pr)

File: config/locales/en.yml [hunk 11]
```diff
@@ -343,6 +343,10 @@ en:
           attributes:
             appropriate_body_id:
               inclusion: Choose an organisation
+        delivery_partners/choose_organisation_form:
+          attributes:
+            delivery_partner_id:
+              inclusion: Choose an organisation
```
Comment (cwrw) [inline config/locales/en.yml:349]: should this be choose a delivery partner 🤔 ?

**Review by cwrw — COMMENTED** [review]
tested and looks great, some minor comments inline

File: config/locales/en.yml (hunk 11 above)
Comment (mooktakim) [reply config/locales/en.yml:349]: @cwrw 
In the UI design we ask them to choose an organisation. The error message should stay consistent

File: app/views/admin/delivery_partners/users/delete.html.erb (hunk 9 above)
Comment (mooktakim) [reply app/views/admin/delivery_partners/users/delete.html.erb:3]: @cwrw 
We're not rendering this page to show the errors. We're deleting and then redirecting to index page

File: app/views/admin/delivery_partners/users/edit.html.erb (hunk 10 above)
Comment (mooktakim) [reply app/views/admin/delivery_partners/users/edit.html.erb:11]: @cwrw 
Added, same for new page

**Review by cwrw — APPROVED** [review]
👍 Thanks for addressing all the comments
